include/nuttx/efuse/efuse.h

```diff
@@ -0,0 +1,155 @@
+ * include/nuttx/efuse/efuse.h
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __INCLUDE_NUTTX_EFUSE_EFUSE_H
+#define __INCLUDE_NUTTX_EFUSE_EFUSE_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <nuttx/fs/ioctl.h>
+
+#include <signal.h>
+#include <stdint.h>
+
+#ifdef CONFIG_EFUSE
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* Command:     EFUSEIOC_READ_FIELD
+ * Description: Read a blob of bits from an efuse field.
+ * Arguments:   A structure containing the field[], a dst pointer, and size
+ *              of bits to be read from efuses.
+ * Return:      Zero (OK) on success.  Minus one will be returned on failure
+ *              with the errno value set appropriately.
+ */
+
+#define EFUSEIOC_READ_FIELD      _EFUSEIOC(0x0001)
+
+/* Command:     EFUSEIOC_WRITE_FIELD
+ * Description: Write a blob of bits to an efuse's field
+ * Arguments:   A structure containing the field[], the src memory and the
+ *              amount of bits to write.
+ * Return:      Zero (OK) on success.  Minus one will be returned on failure
+ *              with the errno value set appropriately.
+ */
+
+#define EFUSEIOC_WRITE_FIELD     _EFUSEIOC(0x0002)
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+/* Structure eFuse field */
+
+struct efuse_desc_s
+{
+  uint16_t   bit_offset; /* Bit offset related to beginning of efuse */
+  uint16_t   bit_count;  /* Length of bit field */
+};
+
+/* Type definition for an eFuse field */
+
+typedef struct efuse_desc_s efuse_desc_t;
+
+/* Structs with the parameters passed to the IOCTLs */
+
+struct efuse_param
+{
+  FAR const efuse_desc_t **field;
+  size_t  size;
+  FAR uint8_t *data;
```

Comment:
can we give an example for data layout? For example, if I have an array:
```
{51, 2},
{60, 4},
```
1. save into two bytes, or into one byte
2. From LSB or MSB if pack into one byte
Need document the detail agreement.

Reply:
struct efuse_param param;
param.field = ESP_EFUSE_CONSOLE_DEBUG_DISABLE;
param.size = 8; /* Number of bits for this efuse field */
param.data = malloc(param.size / 8 + 1);

Reply:
The ESP_EFUSE_CONSOLE_DEBUG_DISABLE is defined this way:
```
static const efuse_desc_t CONSOLE_DEBUG_DISABLE[] =
{
  {
    194, 1   /* Disable ROM BASIC interpreter fallback.
              * EFUSE_RD_CONSOLE_DEBUG_DISABLE
              */
  },
};

const efuse_desc_t *ESP_EFUSE_CONSOLE_DEBUG_DISABLE[] =
{
  &CONSOLE_DEBUG_DISABLE[0],    /* Disable ROM BASIC interpreter fallback.
                                 * EFUSE_RD_CONSOLE_DEBUG_DISABLE.
                                 */
  NULL
};
```

Reply:
@acassis in this example, shouldn't `param.size` be equal to 1 since it is the sum of the sizes of each element in the array?

Reply:
@saramonteiro  The param.size is just a double check to efuse_desc_t bit_size, it is used just to avoid the user pass a wrong ID, but of course we could modify it to be the size of the allocated bytes in param.data if everybody think it is a better idea. @xiaoxiang781216 what do you think?

Reply:
I am fine, but can you demo how two fields(bit size < 8) save into data, it's important to describe the layout clear to avoid the misunderstanding. BTW, please put the description to the header file too.

Reply:
Normally if a field has less than 8 bit, it will be just a single field (no sub-fields), but there is nothing forbidding it to contain for example two sub-fields of 4-bit (or any other combination). In this case the bits will be copied to the 'data' linearly from efuse bit sequence to the 'data' starting at 'data bit 0'.

Reply:
It is flexible because you can "mount" your 'data' with sub-fields pointing to different bits positions and sizes from efuse hardware map, although I never saw it exercised in the IDF code. Normally the use case is just linear efuse bits. I hope you got the idea!

Reply:
Ok, please document the layout, thanks.

Reply:
Ok, thank you!

Reply:
@xiaoxiang781216 I think now all main issues were addressed. I cannot convert **field to *field[] because it needs to be just a point to the ID, the user application will not allocate or create it, but just do i.e.:
param.field = ESP_EFUSE_MAC_CUSTOM_CRC;
Please let me know if everything is fine, if so, please merge it.